File: app/src/main/java/com/sideproject/manlihyang/side/contents/util/Dialog.java
package com.sideproject.manlihyang.side.contents.util;

import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Color;
import android.os.Build;

import androidx.annotation.ColorRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.annotation.StringRes;
import androidx.fragment.app.FragmentManager;

public class Dialog {

    public static MessageDialogFragment showMessage(Context context, FragmentManager fragmentManager, final String message) {
        return new Dialog.Builder(context)
                .setTitle(message)
                .setPositiveTitle("확인")
                .show(fragmentManager);
    }

    public static MessageDialogFragment showMessageConfirm(Context context, FragmentManager fragmentManager, final String message) {
        return new Dialog.Builder(context)
                .setTitle(message)
                .setPositiveTitle("확인")
                .setNegativeTitle("취소")
                .show(fragmentManager);
    }

    public static MessageDialogFragment showMessageConfirm(Context context, FragmentManager fragmentManager, final String message, final String message2) {
        return new Dialog.Builder(context)
                .setTitle(message)
                .setSubTitle(message2)
                .setPositiveTitle("확인")
                .setNegativeTitle("취소")
                .show(fragmentManager);
    }

    public static MessageDialogFragment showSelectRequiredDialogMessage(Context context, FragmentManager fragmentManager, String message) {
        return new Dialog.Builder(context)
                .setTitle(message)
                .setPositiveTitle("확인")
                .setNegativeTitle("취소")
                .show(fragmentManager);
    }

    public static MessageDialogFragment showConfirmDialogMessage(Context context, FragmentManager fragmentManager, String message, String positiveBtnText) {
        return new Dialog.Builder(context)
                .setTitle(message)
                .setNegativeTitle("취소")
                .setPositiveTitle(positiveBtnText)
                .show(fragmentManager);
    }

    public static class Builder{
        @NonNull
        private Context context;
        private int negativeColor;
        private int positiveColor;
        @Nullable
        private String title = null;
        @Nullable
        private String subTitle = null;
        @Nullable
        private String negativeTitle = null;
        @Nullable
        private String positiveTitle = null;
        @Nullable
        private MessageDialogClickListener listener = null;

        public Builder(@NonNull Context context){
            this.context = context;
            this.negativeColor = Color.parseColor("#e5002c");
            this.positiveColor = Color.parseColor("#7304040f");
        }

        public Builder setListener(@Nullable MessageDialogClickListener listener){
            this.listener = listener;
            return this;
        }

        public Builder setTitle(@StringRes int resId){
            this.setTitle(context.getString(resId));
            return this;
        }

        public Builder setSubTitle(@StringRes int resId) {
            this.setSubTitle(context.getString(resId));
            return this;
        }

        public Builder setTitle(@Nullable String title) {
            this.title = title;
            return this;
        }

        public Builder setSubTitle(@Nullable String subTitle) {
            this.subTitle = subTitle;
            return this;
        }

        public Builder setNegativeTitle(@StringRes int resId) {
            this.setNegativeTitle(context.getString(resId));
            return this;
        }

        public Builder setNegativeTitle(@Nullable String negativeTitle) {
            this.negativeTitle = negativeTitle;
            return this;
        }

        @RequiresApi(api = Build.VERSION_CODES.M)
        public Builder setNegativeColor(@ColorRes int resId) {
            this.negativeColor = context.getColor(resId);
            return this;
        }

        @TargetApi(Build.VERSION_CODES.M)
        public Builder setPositiveColor(@ColorRes int positiveColor) {
            this.positiveColor = context.getColor(positiveColor);
            return this;
        }

        public Builder setPositiveTitle(@StringRes int positiveTitle) {
            this.setPositiveTitle(context.getString(positiveTitle));
            return this;
        }

        public Builder setPositiveTitle(@Nullable String positiveTitle) {
            this.positiveTitle = positiveTitle;
            return this;
        }

        public MessageDialogFragment build(){
            final MessageDialogFragment dialogFragment =  MessageDialogFragment.Companion.instantiate(
                    title, subTitle,
                    negativeTitle, positiveTitle,
                    negativeColor, positiveColor
            );
            dialogFragment.setMessageDialogClickListener(listener);
            return dialogFragment;
        }

        public MessageDialogFragment show(@NonNull FragmentManager fragmentManager){
            final MessageDialogFragment dialogFragment = build();
            dialogFragment.show(fragmentManager, "MessageDialogFragment");
            return dialogFragment;
        }
    }
}

File: app/src/main/java/com/sideproject/manlihyang/util/dialog/Dialog.kt
package com.sideproject.manlihyang.util.dialog

import android.content.Context
import androidx.fragment.app.FragmentManager

class Dialog {

    companion object {
        fun showMessage(context: Context, fragmentManager: FragmentManager, message: String, submessage: String = "", extra: Boolean = false, listener: MessageDialogClickListener? = null): MessageDialog {
            return Builder(context)
                .setTitle(message)
                .setSubTitle(submessage)
                .setButtons(extra)
                .setListener(listener)
                .show(fragmentManager)
        }
    }

    class Builder(private val context: Context) {
        private var title: String = ""
        private var subTitle: String = ""
        private var negativeTitle: String = ""
        private var positiveTitle: String = ""
        private var listener: MessageDialogClickListener? = null

        fun setTitle(title: String): Builder {
            this.title = title
            return this
        }

        fun setSubTitle(subTitle: String): Builder {
            this.subTitle = subTitle
            return this
        }

        fun setButtons(extra: Boolean): Builder {
            this.positiveTitle = "확인"
            if(extra) this.negativeTitle = "취소"
            return this
        }

        fun setNegativeTitle(negativeTitle: String): Builder {
            this.negativeTitle = negativeTitle
            return this
        }

        fun setPositiveTitle(positiveTitle: String): Builder {
            this.positiveTitle = positiveTitle
            return this
        }

        fun setListener(listener: MessageDialogClickListener?): Builder {
            listener?.let { this.listener = it }
            return this
        }

        fun build(): MessageDialog {
            return MessageDialog.instantiate(title, subTitle, negativeTitle, positiveTitle).apply {
                listener?.let { setMessageDialogClickListener(it) }
            }
        }

        fun show(fragmentManager: FragmentManager): MessageDialog {
            val dialogFragment = build()
            dialogFragment.show(fragmentManager, "MessageDialog")
            return dialogFragment
        }
    }
}